[concurrency] Implement the Task allocator as bump-pointer allocator.

eeckstein opened the pull request:
The task allocator is implemented by using the new StackAllocator utility.
A StackAllocator performs fast allocation and deallocation of memory by implementing a bump-pointer allocation strategy. In contrast to a pure bump-pointer allocator, it's possible to free memory.
Allocations and deallocations must follow a strict stack discipline.
In general, slabs which become unused are _not_ freed, but reused for subsequent allocations. The first slab is tail allocated to the StackAllocator.

TODO: we could pass an initial pre-allocated first slab to the allocator, which is allocated on the stack or with the parent task's allocator.

rdar://problem/71157018

rxwei commented:
Cool! I was just implementing an allocator in the Swift runtime for use with automatic differentiation — I used `llvm::BumpPtrAllocator` and it worked fine for me. Is there a reason to implement a new allocator instead of using `llvm::BumpPtrAllocator`? If so, perhaps AutoDiff can adopt this implementation as well.

eeckstein commented:
@rxwei There are several differences, most importantly that the task allocator also can free memory (in a stack discipline).
Also, the task allocator uses "inlined" memory for the first slab, which avoid malloc at all if only a small amount of memory is needed.

rxwei commented:
> @rxwei There are several differences, most importantly that the task allocator also can free memory (in a stack discipline).
> Also, the task allocator uses "inlined" memory for the first slab, which avoid malloc at all if only a small amount of memory is needed.

These features would be needed for AutoDiff as well. I’m interested in using this to replace the allocator in #34886. Do you think it can be moved to a header file that we can import?

eeckstein commented:
@swift-ci benchmark

eeckstein committed:
runtime: add a StackAllocator utility

A StackAllocator performs fast allocation and deallocation of memory by implementing a bump-pointer allocation strategy.
In contrast to a pure bump-pointer allocator, it's possible to free memory.
Allocations and deallocations must follow a strict stack discipline.

In general, slabs which become unused are _not_ freed, but reused for subsequent allocations.
The first slab can be placed into pre-allocated memory.

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@rxwei makes sense. I moved it to a separate file.

eeckstein commented on include/swift/ABI/Task.h:
```diff
@@ -199,8 +199,10 @@ class AsyncTask : public HeapObject, public Job {
+  /// Reserved for the use of the task-local stack allocator and it's first
+  /// slab.
+  /// TODO: decide what's the optimal size, e.g. based on typical memory needs.
+  void *AllocatorPrivate[64];
```
Wouldn't every task need at least one context allocation? If so, then this inline-buffer avoids an malloc call.

The size is TBD, though. 64 is just an initial random number.

eeckstein commented:
@swift-ci smoke test linux

eeckstein committed:
[concurrency] Implement the Task allocator as bump-pointer allocator.

Use the StackAllocator as task allocator.

TODO: we could pass an initial pre-allocated first slab to the allocator, which is allocated on the stack or with the parent task's allocator.

rdar://problem/71157018

eeckstein commented on stdlib/public/runtime/StackAllocator.h:
@rjmccall Yes, this is intention. As you said, it's a trade-off.

eeckstein commented:
I pushed a new version, which now does _not_ use a pre-allocated first slab - to keep the task allocation small.
We can improve this in folllow-up work, e.g. by providing stack-allocated space or space allocated with the parent task's allocator.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

rjmccall reviewed (approved):
LGTM
